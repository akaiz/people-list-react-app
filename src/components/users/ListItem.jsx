import React, { useRef, useState } from 'react';
import { useDrag } from 'react-dnd';
import PropTypes from 'prop-types';
import Avatar from 'react-avatar';
import { getDrop } from '../../utils/getDrop';
import UserModalPopUp from './ListItemModal';

function ListItem({
  id, data, index, moveCard, deleteUser
}) {
  const [showModal, SetShowModal] = useState(false);

  function itemDragSettings() {
    const ref = useRef(null);
    const [, drop] = getDrop(ref, index, moveCard, 'User');
    const [{ isDragging }, drag] = useDrag({
      item: { type: 'User', id, index },
      collect: (monitor) => ({
        isDragging: monitor.isDragging(),
      }),
    });
    const opacity = isDragging ? 0 : 1;
    drag(drop(ref));
    return { ref, opacity };
  }
  const { ref, opacity } = itemDragSettings();

  const itemOnClick = () => {
    SetShowModal(!showModal);
  };
  const picture = data.picture_id ? data.picture_id.pictures[128] : null;
  return (
    <div
      className="row user-item"
      ref={ref}
      style={{ opacity }}
      onClick={itemOnClick}
      onKeyDown={itemOnClick}
      role="presentation"
      data-test="user-item"
    >
      <div className="col-xs-5 col-sm-10 col-lg-10 col-xl-11 user-item-col-1">
        <h4 data-test="user-name">{data.name}</h4>
        <h6 data-test="user-org"><span><i className="fa fa-building-o" /></span> {data.org_name}</h6>
      </div>
      <div className="col-xs-3 col-sm-2 col-lg-2 col-xl-1" data-test="user-photo">
        <Avatar src={picture} round="50px" size="100" fgColor="#357edd" name={data.name} color="#d1dff0" />
      </div>
      <UserModalPopUp
        data={{ ...data, user_photo: picture }}
        showModal={showModal}
        deleteUser={deleteUser}
        toggleModal={(value) => SetShowModal(value)}
      />
    </div>
  );
}

ListItem.propTypes = {
  data: PropTypes.object.isRequired,
  index: PropTypes.number.isRequired,
  id: PropTypes.number.isRequired,
  moveCard: PropTypes.func,
  deleteUser: PropTypes.func,
};
export default ListItem;
