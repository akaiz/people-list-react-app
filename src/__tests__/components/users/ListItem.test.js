import React from 'react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { DndProvider } from 'react-dnd';
import HTML5Backend from 'react-dnd-html5-backend';
import Avatar from 'react-avatar';
import renderer from 'react-test-renderer';
import ListItem from '../../../components/users/ListItem';

import { apiResponse, findByAttr } from '../../TestHelper';

configure({ adapter: new Adapter() });

const setUp = (props = {}) => {
  const component = mount(
    <DndProvider backend={HTML5Backend}>
      <ListItem {...props} />
    </DndProvider>);
  return component;
};


describe('<ListItem> Tests', () => {
  let wrapper;
  const user = apiResponse(1).data[0];
  const props = {
    data: user, index: 1, id: 1, deleteUser: jest.fn(() => {})
  };
  beforeEach(() => {
    wrapper = setUp(props);
  });

  describe('Checking Props', () => {
    it('Should receive the sent props', () => {
      expect(wrapper.find(ListItem).props()).toStrictEqual(props);
    });
  });

  describe('Check rendered details', () => {
    it('Should render user name ', () => {
      expect(findByAttr(wrapper.find(ListItem), 'user-name').text()).toBe(props.data.name);
    });

    it('Should render user organization ', () => {
      expect(findByAttr(wrapper.find(ListItem), 'user-org').text().trim()).toMatch(user.org_name);
    });

    it('Should render user photo ', () => {
      expect(wrapper.find(ListItem).find(Avatar).length).toBe(1);
    });
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <DndProvider backend={HTML5Backend}>
          <ListItem {...props} />
        </DndProvider>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
