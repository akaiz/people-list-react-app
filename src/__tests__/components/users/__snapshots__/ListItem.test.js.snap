// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListItem> Tests renders correctly 1`] = `
<div
  className="row user-item"
  data-test="user-item"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="presentation"
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <div
    className="col-xs-5 col-sm-10 col-lg-10 col-xl-11 user-item-col-1"
  >
    <h4
      data-test="user-name"
    >
      Name0
    </h4>
    <h6
      data-test="user-org"
    >
      <span>
        <i
          className="fa fa-building-o"
        />
      </span>
       
      Org name 0
    </h6>
  </div>
  <div
    className="col-xs-3 col-sm-2 col-lg-2 col-xl-1"
    data-test="user-photo"
  >
    <div
      className=" sb-avatar"
      style={
        Object {
          "borderRadius": "50px",
          "display": "inline-block",
          "fontFamily": "Helvetica, Arial, sans-serif",
          "height": "100px",
          "verticalAlign": "middle",
          "width": "100px",
        }
      }
    >
      <div
        className=" sb-avatar__text"
        style={
          Object {
            "background": "#d1dff0",
            "borderRadius": "50px",
            "color": "#357edd",
            "height": "100px",
            "lineHeight": "initial",
            "textAlign": "center",
            "textTransform": "uppercase",
            "width": "100px",
          }
        }
      >
        <div
          style={
            Object {
              "display": "table",
              "height": "100%",
              "tableLayout": "fixed",
              "width": "100%",
            }
          }
        >
          <span
            style={
              Object {
                "display": "table-cell",
                "fontSize": "100%",
                "verticalAlign": "middle",
                "whiteSpace": "nowrap",
              }
            }
          >
            <span />
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
